### PR TITLE
[RFC] Symfony RememberMe Service Implementation

### PR DESCRIPTION
**Pull request by DanielSchwiperich**
This is the PR for Issue #454 

The current status is completely working but I need Feedback before going further
## What is this PR for:

Providing autologin functionality across different devices for a frontend user. The current implementation only allows autologin for one device.
## Questions
1. **Make the Implementation optional, parallel or replace?** We could add _Allow autologin from multiple devices_ to tl_settings. See https://github.com/contao/core-bundle/issues/454#issuecomment-217706318 for infos on it.  
   I would vote for optional in combination with parallel for the most user comfort with deprecation warning for Contao 5. (Switching from normal autologin to multi-autologin would not break autologin cookies)
2. **How to handle the _cookie lifetime_ param?** Since the RememberMeService needs this setting at startup and cannot be changed later I had to put it in the DI Container for now. There the `$GLOBALS['TL_CONFIG']['autologin']` is not available atm. So we either had to 
   - remove it from tl_settings
   - write it back to the parameters.yml (bad practice?) 
   - load the localconfig.php with default.php (as fallback) in DI / CoreExtension (also bad practice?)
   - write it to another, dynamic configuration.yml and load in  DI / CoreExtension
## Further Implementation Infos
- Currently the rememberMeService lives beside the current autologin implementation (which is no problem, they do not harm each other in any way)
- The contao.logger you see in the services.yml / logger class is for development. The Symfony rememberMe Service class can _only_ handle a psr-3 compatible logger, so I made a quick wrapper for it. Will be removed 
- I had to extend the Contao UserProvider to make it compatible to the remember service. The current implementation does not allow to really load a user. I'm sure this was done on purpose by @aschempp but don't know why
- The rememberMeService requires The User Object to implement the `UserInterface`. I moved that part to the `User` class instead of the `FrontendUser`, so this has not to be changed for future auth implementations where the backend is also involved.
- The authenticate method diff looks bigger than it is. I had to move one part and restructure a little bit to avoid code duplication
- Code is not very pretty atm. Are there any Code / Style Conventions? Didn't found any info on it
## How To test
- Setup a Contao Installation with a login form and autologin enabled
- Create a member user
- Login to the frontend
- Delete the FE_AUTH and FE_AUTOLOGIN cookie, reload the site (currently the rememberMeService uses it's own Cookie FE_AUTOLOGIN_MULTI). See the user should still be logged in. 
## Todos
- [ ] Write Unittests
- [x] Implementation of Question 1,2
- [ ] Code Style

---

~~I had a to rush a little bit when I was half through the PR, because guests arrived for BQ, but I wanted to get out to you. Will reread that later and _optimize_~~ <- done :smile:

### CODE REVIEW

**Comment by discordier** [comment]
> Make the Implementation optional, parallel or replace?

I also vote for optional here.

> How to handle the cookie lifetime param?

IMO we should move it to parameters, we already agreed in the past that we want to move everything from `localconfig.php` to `parameters.yml` eventually when we are touching the area where it applies.
The way of accessing config values via `$GLOBALS['TL_CONFIG']` is not supported in Contao 4 and all access must be done either via DI or via `\Config::get()` calls. Therefore it is IMO totally fine to push it to a config parameter.

For your additions to the user classes and the user provider.
I wonder if it now would make sense to implement a real user provider, user class (implementing the UserInterface) and Authenticator (implementing the current authenticate functionality) and only delegate from the legacy classes.
This would allow us to use all the authentication mechanisms of Symfony.
Until now there was no need to implement something like this but as you have to alter the current implementation anyway, maybe it is time to do so. IMO the implementation of this PR will be slightly slimmer then, as it does not need to change implementation (aside from activating).

On your general question about code style, we are mostly using the symfony code style for new code and the "Contao legacy style" in legacy code.

**Comment by DanielSchwiperich** [comment]
Thanks for feedback @discordier 
Ok I'll continue and make it optional.

I agree with you that the authentication system needs an overhaul. Also regardings stuff like auth providers, events instead of hooks ( + more events like fire an event if a user gets blocked) etc. 
But please not **now** and not in **this** PR :wink: (This would be something for Contao 5 I think, since it would be hard to not break BC)
My additions to the `UserProvider` and the `User` are pretty simple / trivial. Would like to hear if @aschempp  is ok with it, since he implemented it originally (at least that's the doc block saying, didn't check history)

**Comment by DanielSchwiperich** [comment]
I made the remember_me service optional and by default off. Can be set in the backend via  the security panel.

What about translations. I don't need to provide all, do I? :fearful:  

I'm not so happy to have the expiration value set in the parameters.yml / DI for the multi-autologin session and in tl_settings for the single-autologin sessions. 
This does not feel right. I think they should always be the same. But we cannot remove it from tl_settings without breaking BC. And I cannot use the value from tl_settings since the it is not available on container / DI  setup in `ContaoCoreExtension::loadInternal` :disappointed: 

Will do the Unittests on the weekend probably.

**Comment by aschempp** [comment]
Translations are usually managed on Transifex? 

I think we need to move it to the container, we did the same for several other parameters. It will be BC in the sense that we write container parameters to `$GLOBALS['TL_CONFIG']`, but if adjusted the setting in `localconfig.php` will not be applied for the multi-login, which is fine for me.

We're only adding things to the parameters.yml that must be adjusted on each installations. Things that are usually the same can simply be bundle config options with default values (see `url_suffix` for example).

**Comment by DanielSchwiperich** [comment]
yeah Transifex is a chapter I completely missed / skipped until now. Did take a look, still not sure if it's ok / enough whats in the PR or if it's even wrong?

Regarding the expiration value, I'll replace the current `tl_setting` / `$GLOBALS['TL_CONFIG']`  autologin value then in this PR like done with `url_suffix`. 
thanks for feedback

**Comment by DanielSchwiperich** [comment]
So the `autologin` setting has gone from the backned and will be autofilled with the container param. 

I've to say it works really well so far. 
Only Unittests are missing now. 

@leofeyer Do you want me to use the `Config` class from global Namespace to be on par with the rest of the class, leave it combined, or should I adjust the other usages in this class, since it's now now BC anymore to contao 3?

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -99,6 +99,23 @@ public function getConfigTreeBuilder()
...
+                            ->children()
+                                ->scalarNode('cookie_name')
+                                    ->cannotBeEmpty()
+                                    ->defaultValue('FE_AUTO_LOGIN_MULTI')
+                                ->end()
+                                ->integerNode('cookie_lifetime')
+                                    ->min(0)
+                                    ->defaultValue(7776000)
+                                ->end()
+                                ->scalarNode('formfield_name')
```
Comment (leofeyer) [inline src/DependencyInjection/Configuration.php]: Why does this need to be configurable?

**Comment by leofeyer** [comment]
Why did you make the new feature optional? We should not provide a flag to enable or disable it but simply replace the existing routines.

**Comment by leofeyer** [comment]
Haha, I just realised that this was one of your initial questions. 😄 

> Make the Implementation optional, parallel or replace?

Replace. Definitely.

> How to handle the cookie lifetime param?

A container parameter seems fine to me.

File: src/DependencyInjection/Configuration.php (hunk 1 above)
Comment (aschempp) [inline src/DependencyInjection/Configuration.php]: Probably just best practice because Symfony allows to configure field names of the form login as well.

**Comment by aschempp** [comment]
this probably conflicts with #559

**Comment by bytehead** [comment]
I don't know if it there are some future problems, if you merge this PR and then I try to adjust/implement this changes to #559?

**Comment by leofeyer** [comment]
We should definitely consider #559. But we should also merge this PR, shouldn't we? 😄

**Comment by DanielSchwiperich** [comment]
I alreaday mentioned in #559 that there may some overlapping. 
I choosed to make the feature optional because @discordier voted for this on my question (and it was the only answer at that point 😉 ) 
But sure I can change that. I was for optional because it changes the behaviour and theoretically people could rely on the current one. I have no idea why anyone only wants to have autologin just for one device nut who knows. So it depends on if you consider the current implementation as *bugged* or  not. You say 😸 

Just let me know know if I can / should finish the implementation (unit tests are also missing right now and I have to adjust compatibility with current release). Will be happy to do so.

**Comment by leofeyer** [comment]
Definite +1 from me!

**Comment by leofeyer** [comment]
@contao/developers What about you?

**Comment by Toflar** [comment]
I definitely like to see a remember me setting! But also, I like to see the whole auth being processed via Symfony so we can push this further and add OAuth etc. Unfortunately, I cannot invest my time here because there's too much other stuff I want to work on. So in the best case @DanielSchwiperich and @bytehead can come up with a complete PR that handles the BE and FE auth via Symfony and provides a remember me function too :)

**Comment by leofeyer** [comment]
As discussed at the Contao Camp:

* We want to replace the current implementation instead of having it optional.
* We do not want to add any configuration parameters. Things should be read from the Sf configuration or the default values.

The only remaining conflict is that this PR makes the `FrontendUser` class implement the `UserInterface` and #559 makes the `UserModel` class implement the `UserInterface`.

File: src/Resources/contao/library/Contao/User.php [hunk 2]
```diff
@@ -673,6 +675,47 @@ public function isMemberOf($id)
 		return false;
 	}
 
+	/**
+	 * @inheritDoc
+	 */
+	public function getRoles()
+	{
+		return [];
```
Comment (aschempp) [inline src/Resources/contao/library/Contao/User.php]: This should be the same roles as defined in the Token

**Comment by aschempp** [comment]
So yesterday I took a few hours to look into this, as we wanted to merge it into Contao 4.3 as a bug fix. I fixed several little issues (like removing the config options), and then digged again into Symfony security. I re-found what I implemented two years ago (for #559), and unfortunately also found issues with this implementation that cannot be easily fixed.

As you know, this implementation no longer requires a database table to hold the identification key, as Contao currently does with `tl_member.autologin`. The authentication is solely based on the cookie which is encrypted with the `kernel.secret`. The issue with this is, if someone steals your cookie, there is no way to revoke the cookie.

In the current implementation, a device is automatically *revoked* if another device is authenticated. This would no longer be possible, every device that once has that cookie will login forever. I don't think that's the solution we want, as it would lower security of the application.

There is a solution build into Symfony. Instead of using the `TokenBasedRememberMeService` we would need to use the `PersistentTokenBasedRememberMeServices`. That requires a new database table though, and implementation would not be done in a few hours.

A good read on the topic of *persistent tokens* can be found at https://paragonie.com/blog/2015/04/secure-authentication-php-with-long-term-persistence
A quick summary is at http://stackoverflow.com/questions/244882/what-is-the-best-way-to-implement-remember-me-for-a-website#answer-244907

**Comment by leofeyer** [comment]
> Instead of using the `TokenBasedRememberMeService` we would need to use the `PersistentTokenBasedRememberMeServices`. That requires a new database table though, and implementation would not be done in a few hours.

Still I think this is the right way to go. The current implementation is not secure (enough).

**Comment by aschempp** [comment]
Absolutely, but not in 4.3 rc imho...

**Comment by leofeyer** [comment]
Since #685 is almost ready to merge, I'm closing this PR in favor of #685.

**Pull request closed without merging**